### DISCUSSION TITLE
how to build docker?

### DISCUSSION THREAD

**Discussion by asnaeb** [vercel/next.js#52466] [App Router] [answered]
### Summary

The [with-docker](https://github.com/vercel/next.js/tree/canary/examples/with-docker) example only shows how to build for `pages` dir but I haven't found and info on how to build for `app` dir. My project uses `src/app`.

I have tried using the following Dockerfile (adapted from the example)

```dockerfile
FROM node:18-alpine AS base

FROM base AS deps
RUN apk add --no-cache libc6-compat
WORKDIR /app

COPY package.json package-lock.json* ./
RUN npm ci

FROM base AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .

ENV NEXT_TELEMETRY_DISABLED 1

RUN npm run build

FROM base AS runner
WORKDIR /app

ENV NODE_ENV production
ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/public ./public

COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs

EXPOSE 3000

ENV PORT 3000

CMD ["node", "server.js"]
```
with the following next.config.js
```javascript
const nextConfig = {
    output: "standalone"
};
```
It gives the following error when executing `npm run build` at `builder` stage
```
Couldn't find any `pages` or `app` directory. Please create one under the project root
```
So I have tried creating an empty `pages` directory and I got it to build but when running it, I only get `404`s for every route. I guess that's because It build thinking I was using `pages` instead of `src/app`. Is there any official way to make it work using appDir or am I doing something wrong here?

### Additional information

My project is using `app` dir with `src` dir. So the relevant files are located in `src/app`. It runs ok with `npm run build` - `npm start`.

**Comment by icyJoseph** [comment]
Hi,

You must be doing something wrong. I have created a with-docker example, and without changing anything in the Dockerfile, nor next.config.js, I just deleted pages, created `src/app`, and added a layout and page files. Run docker build and it worked.

This is my file structure:

```console
.
├── Dockerfile
├── README.md
├── next.config.js
├── package.json
├── public
│   ├── favicon.ico
│   └── vercel.svg
├── src
│   ├── app
│   │   ├── layout.js
│   │   └── page.js
│   └── styles
│       ├── Home.module.css
│       └── globals.css
└── yarn.lock
```

Could you share a bit more about your setup?

**Reply by asnaeb** [reply]
<img width="303" alt="Screenshot 2023-07-10 at 09 42 32" src="https://github.com/vercel/next.js/assets/117033917/283da941-4016-43fd-9c0a-71e5d24a49d2">

This is my project structure. I am using npm in place of yarn and I have tried using the Dockerfile and next.config.js from the example but I always get the same result:
```
 Couldn't find any `pages` or `app` directory. Please create one under the project root
```

**Reply by icyJoseph** [reply]
What's in `dockerignore`?

**Reply by asnaeb** [reply, marked as answer]
ok. It's my fault. I had `src` in dockerignore from a different branch. Thank you for thinking about it.

**Reply by icyJoseph** [reply]
Alright, awesome! Don't forget to mark as answered and close it :)
